Add JarLauncherDetector main class

Kehrlann opened the pull request:
## Reflectively detect which `JarLauncher` to use

Closes gh-32

Open questions:
- [x] How to test? Currently doing unit-tests with included jars, but that adds 40MB
- [x] What to do no `JarLauncher` is launched? Currently throwing a RuntimeException with little info.

rwinch commented:
@Kehrlann Thanks for the draft PR

We spoke about the testing and I think that mocking the static method makes sense.

As for if no JarLauncher is found, I'd add the SpringBootApplicationMain as the fallback to JarLauncherDetector.

I'd also make the default main class the new `JarLauncherDetector`

Kehrlann commented:
@rwinch addressed comments.

I'm still wondering whether we should log something when the `JarLauncherDetector` runs and fails to find a suitable launcher, warning the user that we're falling back on the default `SpringBootApplicationMain`.

Since the Spring app is not running yet, slf4j is not bootstrapped, so I'm unsure what to do.

rwinch commented:
Thanks for the Pull Request! This is now merged into main :smile:

rwinch commented:
@Kehrlann Thank you this looks good. Let's handle the concerns you had around failures in a separate ticket.
